dev/fattest.simplicity/src/componenttest/containers/ImageBuilder.java

```diff
@@ -0,0 +1,232 @@
+import java.util.Objects;
+import java.util.Optional;
+import java.util.stream.Stream;
+
+import org.testcontainers.images.PullPolicy;
+import org.testcontainers.images.RemoteDockerImage;
+import org.testcontainers.images.builder.ImageFromDockerfile;
+import org.testcontainers.utility.DockerImageName;
+import org.testcontainers.utility.ImageNameSubstitutor;
+import org.testcontainers.utility.MountableFile;
+
+import com.ibm.websphere.simplicity.log.Log;
+
+/**
+ * This builder class is an extension of {@link org.testcontainers.images.builder.ImageFromDockerfile}
+ * and is intended to allow developers of Open Liberty the ability to create custom images from a Dockerfile.
+ *
+ * This class will is optimized to find or build the image using the following priority:
+ * - If the image is cached on the docker host, use it.
+ * - Otherwise, if the image is cached in the registry, pull it, and use it.
+ * - Otherwise, build the image and cache it on the docker host.
+ *
+ * TODO make builder public once build pipeline is finalized
+ */
+class ImageBuilder {
+
+    private static final Class<?> c = ImageBuilder.class;
+
+    // The --build-arg necessary to overwrite the default BASE_IMAGE in the Dockerfile
+    // with the mirrored image in artifactory
+    private static final String BASE_IMAGE = "BASE_IMAGE";
+
+    // Image to build
+    private final DockerImageName image;
+
+    // Constructor
+    private ImageBuilder(DockerImageName image) {
+        //builder class
+        this.image = image;
+    }
+
+    /**
+     * The image to build.
+     *
+     * The Dockerfile with instructions on how to build this image must be saved in source control in directory
+     * io.openliberty.org.testcontainers/resources/openliberty/testcontainers/<image-name>/<image-version>/Dockerfile
+     *
+     * Note: The resulting image will be cached with name "localhost/openliberty/testcontainers/<image-name>:<image-version>"
+     * therefore, you must update the image version whenever a change is made to the corresponding Dockerfile.
+     *
+     * @param  customImage the image to build in format "<image-name>:<image-version>" or "openliberty/testcontainers/<image-name>:<image-version>"
+     *
+     * @return             instance of ImageBuilder
+     */
+    public static ImageBuilder build(String customImage) {
+        Objects.requireNonNull(customImage);
+
+        return new ImageBuilder(ImageBuilderSubstitutor.instance().apply(DockerImageName.parse(customImage)));
+    }
+
+    // Add future configuration methods here
+
+    /**
+     * Termination point of this builder class.
+     *
+     * We will first attempt to find a cached version of the image,
+     * if unsuccessful, we will then build the image from the Dockerfile.
+     *
+     * @return RemoteDockerImage that points to a cached or built image.
+     */
+    public RemoteDockerImage get() {
+        return getCached().orElse(pullCached().orElse(buildFromDockerfile()));
```

Comment:
I was looking into orElse() and I see [this](https://www.baeldung.com/java-optional-or-else-vs-or-else-get):
> With it, we can easily infer that the parameter of orElse() is evaluated, even when having a non-empty Optional.
If I understand that correctly, even if `getCached()` returns a value, we will still end up calling `pullCached()` and `buildFromDockerfile()`

Reply:
Great catch, updated!